rocketpy/mathutils/function.py

```diff
@@ -2108,61 +2130,60 @@ def __add__(self, other):  # pylint: disable=too-many-statements
+            return Function(source, inputs, outputs, interp, extrap)
+        elif isinstance(other, NUMERICAL_TYPES) or self.__is_single_element_array(
+            other
+        ):
+            if self._source_type is SourceType.ARRAY:
+                source = np.column_stack((self._domain, np.add(self._image, other)))
+                outputs = f"({self.__outputs__[0]}+{other})"
+                return Function(source, inputs, outputs, interp, extrap)
+            else:
+                return Function(
+                    self.__make_arith_lambda(
+                        operator.add, self.get_value_opt, other, dom_dim
+                    ),
+                    inputs,
+                )
+        elif callable(other):
+            if other_is_func:
+                other_dim = other.__dom_dim__
+                other = other.get_value_opt if other_is_array else other.source
             else:
-                return Function(lambda x: (self.get_value_opt(x) + other(x)))
-        # If other is Float except...
-        except AttributeError:
-            if isinstance(other, NUMERICAL_TYPES) or self.__is_single_element_array(
-                other
-            ):
-                # Check if Function object source is array or callable
-                if isinstance(self.source, np.ndarray):
-                    # Operate on grid values
-                    ys = self.y_array + other
-                    xs = self.x_array
-                    source = np.concatenate(([xs], [ys])).transpose()
-                    # Retrieve inputs, outputs and interpolation
-                    inputs = self.__inputs__[:]
-                    outputs = self.__outputs__[0] + " + " + str(other)
-                    outputs = "(" + outputs + ")"
-                    interpolation = self.__interpolation__
-                    extrapolation = self.__extrapolation__
-                    # Create new Function object
-                    return Function(
-                        source, inputs, outputs, interpolation, extrapolation
+                other_dim = len(signature(other).parameters)
+
+            if dom_dim == 1 or other_dim == 1 or dom_dim == other_dim:
+                return Function(
+                    self.__make_arith_lambda(
+                        operator.add, self.get_value_opt, other, dom_dim, other_dim
                     )
-                else:
-                    return Function(lambda x: (self.get_value_opt(x) + other))
-            # Or if it is just a callable
-            elif callable(other):
-                return Function(lambda x: (self.get_value_opt(x) + other(x)))
+                )
+            else:
+                raise TypeError(
+                    f"The number of parameters in the function to be added ({other_dim}) "
+                    f"does not match the number of parameters of the Function ({dom_dim})."
+                )
+        raise TypeError("Unsupported type for addition")
```

Comment:
I usually don't request these kind of changes, but I believe in this case it's worth it, since it regards to the Function class.

Codecoverage is reporting that you didn't cover the raise exceptions. Please add more tests or skip the rows using `# pragma: no-cover`